src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+    general_purpose_fuse_register_lo: VolatileCell<u32>,
+}
+
+const FLASHCALW_BASE_ADDRS : *mut Registers = 0x400A0000 as *mut Registers;
+
+#[allow(dead_code)]
+enum RegKey {
+    CONTROL,
+    COMMAND,
+    STATUS,
+    PARAMETER,
+    VERSION,
+    GPFRHI,
+    GPFRLO
+}
+
+/// Error codes are used to inform the Client if the command completed successfully
+/// or whether there was an error and what type of error it was.
+pub enum Error {
+    CommandComplete,    /* Command Complete */
+    LockE,              /* Lock Error (i.e. tried writing to locked page) */
+    ProgE,              /* Program Error (i.e. incorrectly issued flash commands */
+    LockProgE,          /* Lock and Program Error */
+    ECC,                /* Error Correcting Code Error */
+}
+
+/// High level commands to issue to the flash. Usually to track the state of 
+/// a command especially if it's multiple FlashCMDs. 
+/// 
+/// For example an erase is: 1) Unlock Page  (UP)
+///                          2) Erase Page   (EP)
+///                          3) Lock Page    (LP)
+/// Store what high level command we're doing allows us to track the state and
+/// continue the steps of the command in handle_interrupt.
+#[derive(Clone, Copy, PartialEq)]
+pub enum Command {
+    Write,
+    Read,
+    Erase,
+    None
+}
+
+
+
+/// There are 18 recognized commands for the flash. These are 'bare-bones' commands
+/// and values that are written to the Flash's command register to inform 
+/// the flash what to do. Table 14-5.
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashCMD {
+    NOP,
+    WP,
+    EP,
+    CPB,
+    LP,
+    UP,
+    EA,
+    WGPB,
+    EGPB,
+    SSB,
+    PGPFB,
+    EAGPF,
+    QPR,
+    WUP,
+    EUP,
+    QPRUP,
+    HSEN,
+    HSDIS,
+}
+
+//The two Flash speeds
+#[derive(Clone, Copy)]
+pub enum Speed {
+    Standard,
+    HighSpeed
+}
+
+/// FlashState is used to track the current state of the flash in high level
+/// command.
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashState {
+    Locking,
+    Unlocking,
+    WritePageBuffer,
+    Writing,
+    Reading,
+    Erasing,
+    Ready,
+    Unconfigured
+}
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
+    registers: *mut Registers,
+    ahb_clock: pm::Clock,
+    hramc1_clock: pm::Clock,
+    pb_clock: pm::Clock,
+    wait_until_ready: fn(&FLASHCALW) -> (),
+    error_status: TakeCell<u32>,
+    ready: TakeCell<bool>,
+    client: TakeCell<&'static Client>,
+    current_state: Cell<FlashState>,
+    current_command: Cell<Command>,
+    page: Cell<i32>,
+    page_buffer: TakeCell<[u8; FLASH_PAGE_SIZE as usize]>
+}
+
+//static instance for the board. Only one FLASHCALW on chip.
+pub static mut flash_controller : FLASHCALW = 
+    FLASHCALW::new(FLASHCALW_BASE_ADDRS, pm::HSBClock::FLASHCALW, 
+        pm::HSBClock::FLASHCALWP, pm::PBBClock::FLASHCALW);
+
+
+// Few constants relating to module configuration.
+const FLASH_PAGE_SIZE : u32 = 512;
```

Comment:
For all these constants, i think it's clearer without the `FLASH_` prefix. We're already in the flashcalw module, so we know what they are related too. `FLASH_` style prefixes are a relic from C, which doesn't have a namespace system.